### PR TITLE
#161967129 create user profile

### PR DESCRIPTION
**Pull request by denomash**
**What does this PR do?**
- creates a profile user upon registration

**Description of Task to be completed?**
- users can view their profiles
- users can view other users profile
- users can edit their own profile
- users can not edit other users profile

**How should this be manually tested?**

- Clone the repo and cd into the project folder
- Create a virtual environment by running `python3 -m venv <name_of_environment>`
- Activate the virtual environment by running `source <name_of_environment>/bin/activate`
- Run `$ pip install -r requirements.txt`
- Then add the following environment variables to your computer:

 - `$ export DB_NAME='choose_a_dbname'`
 - `$ export DB_USER='db_username'`
 - `$ export DB_PASSWORD='db_role_password'`
 - `$ export DB_HOST='localhost or preffered host'`
 - `$ export DB_PORT='server_port_number'`
- `$ export SECRET_KEY='some_secret_key`
- `$ export SOCIAL_AUTH_FACEBOOK_KEY='<facebook_app_id>'`
- `$ export SOCIAL_AUTH_FACEBOOK_SECRET='<facebook_app_secret>'`
- `$ export SOCIAL_AUTH_TWITTER_KEY='<twitter_api_key>'`
- `$ export SOCIAL_AUTH_TWITTER_SECRET='<twitter_api_secret>'`
- `$ export SOCIAL_AUTH_GOOGLE_OAUTH2_KEY='<google_client_id>'`
- `$ export SOCIAL_AUTH_GOOGLE_OAUTH2_SECRET='<google_client_secret>'`

- Change the current branch by running `git checkout  ft-create-user-profile-161967129`
- Run `python3 manage.py makemigrations` then, `python3 manage.py migrate`
- Finally run `python3 manage.py runserver`
- To view your profile or others profile, use the following endpoint: `http://127.0.0.1:8000/profiles/<username>`
- To edit your profile , use the following endpoint: `http://127.0.0.1:8000/profiles/edit/<username>`

**What are the relevant pivotal tracker stories?**
- [#161967129](https://www.pivotaltracker.com/story/show/161967129/comments/197477151)

**Checklist (edited)**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

**Screenshots (if appropriate)**
**A USER CAN VIEW THEIR OWN PROFILE**
screenshot 2018-12-13 at 11 23 46

**A USER CAN VIEW OTHER USER'S PROFILE**
screenshot 2018-12-13 at 11 33 36

**A USER CAN EDIT THEIR OWN PROFILE**
screenshot 2018-12-13 at 11 26 07

**A USER CAN NOT EDIT OTHER USERS PROFILE**
screenshot 2018-12-11 at 11 45 48

### CODE REVIEW

File: authors/apps/profiles/urls.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+
+from django.urls import path
+
+
+# local imports
+from .views import GetUserProfile, UpdateUserProfile
+
+
+urlpatterns = [
+    path('profiles/<username>', GetUserProfile.as_view(), name='profile'),
```
Comment (Darius-Ndubi) [inline authors/apps/profiles/urls.py]: Be specific on what should be entered like/<str:username>

**Review by jillo-abdullahi — CHANGES_REQUESTED** [review]
For testing purposes please add the setup configurations such as registering a user, activating a user, and log in to a setup module so that other tests can just do an import instead of recreating the setup process.

**Review by Darius-Ndubi — CHANGES_REQUESTED** [review]

File: authors/apps/profiles/models.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+from django.dispatch import receiver
+
+
+class Profile(models.Model):
+    """
+    create a user profile model
+    """
+
+    user = models.OneToOneField(User, on_delete=models.CASCADE)
+    username = models.CharField(max_length=20, blank=False)
```
Comment (bettblake08) [inline authors/apps/profiles/models.py]: I think having a username in two models referring to one user is redundant. Can you find a way to remove this?

File: authors/apps/profiles/urls.py (hunk 1 above)
Comment (denomash) [reply authors/apps/profiles/urls.py]: noted. let me look into that

**Review by bettblake08 — CHANGES_REQUESTED** [review]

**Comment by denomash** [comment]
@jillo-abdullahi that's a good suggestion with regard to testing purposes. let me look into it

File: authors/apps/profiles/models.py (hunk 2 above)
Comment (denomash) [reply authors/apps/profiles/models.py]: I am using the username field to query the profile model while getting and updating the profile model. Let me check if there is another way

**Review by jillo-abdullahi — APPROVED** [review]

File: authors/apps/profiles/urls.py (hunk 1 above)
Comment (denomash) [reply authors/apps/profiles/urls.py]: made the changes

**Comment by denomash** [comment]
@jillo-abdullahi it's all good. I made the changes

**Review by Darius-Ndubi — APPROVED** [review]

**Review by Bryan-Cee — APPROVED** [review]

File: authors/apps/profiles/models.py (hunk 2 above)
Comment (mutoromike) [reply authors/apps/profiles/models.py]: Can you use a fkey to have it point to the same user in the User model?

**Review by Bryan-Cee — CHANGES_REQUESTED** [review]
Please make sure your return response conforms with what is in the readme, check on your serializer.
screenshot 2018-12-12 at 12 07 32

The response I get when I want to view a profile is this:
screenshot 2018-12-12 at 12 07 48

File: authors/apps/profiles/models.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    bio = models.TextField(blank=True)
+    image = models.URLField(
+        blank=True,
+        default="https://d1nhio0ox7pgb.cloudfront.net/_img/o_collection_png/green_dark_grey/512x512/plain/user.png"
+    )
+    company = models.CharField(max_length=100, blank=True)
```
Comment (Bryan-Cee) [inline authors/apps/profiles/models.py:19]: Regarding the field that have been added, I believe we should consult with the PO and the TTL to clarify what is required.

File: authors/apps/profiles/urls.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+
+from django.urls import path
+
+
+# local imports
+from .views import GetUserProfile, UpdateUserProfile
```
Comment (bettblake08) [inline authors/apps/profiles/urls.py]: Can you change the name of the views to add View to the end, to conform to the other view?

**Review by bettblake08 — CHANGES_REQUESTED** [review]

File: authors/apps/profiles/urls.py (hunk 4 above)
Comment (denomash) [reply authors/apps/profiles/urls.py]: Kindly elaborate on this.

Comment (jillo-abdullahi) [reply authors/apps/profiles/urls.py]: I think what @bettblake08 means is that you use something like 'GetUserProfileView' or 'UpdateUserProfileView'

Comment (denomash) [reply authors/apps/profiles/urls.py]: @jillo-abdullahi, got it. Thanks for clearing the air on that.

Comment (jillo-abdullahi) [reply authors/apps/profiles/urls.py]: 👍

File: authors/apps/profiles/models.py (hunk 3 above)
Comment (denomash) [reply authors/apps/profiles/models.py:19]: With regard to the following field, I believe that was your task. And about the fields thats where creativity came in as described in the task. But I agree with you on consulting with the PO and the TTL to clarify what is required.

Comment (denomash) [reply authors/apps/profiles/models.py:19]: @Bryan-Cee, at the end of it all that is what profile is supposed to be like.

**Review by bettblake08 — APPROVED** [review]
👍 Looks good

**Review by Bryan-Cee — APPROVED** [review]

**Review by mutoromike — APPROVED** [review]
Good work increasing test coverage!

**Pull request merged**